### PR TITLE
Initial Segmentation Modeling Framework

### PR DESCRIPTION
**Pull request by MattsonCam**
Includes the framework for training the first segmentation models. Many compromises were made to ensure extensibility and reproducibility while minimizing the time required for development. Therefore, this framework will likely change if model development persists.

### CODE REVIEW

File: datasets/dataset_00/utils/ImageSelector.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        image_height,
+        image_width,
+        **kwargs,
+    ) -> None:
+        self.input_max_pixel_value = input_max_pixel_value
+        self.image_height = image_height
+        self.image_width = image_width
+
+    def select_zslices(
+        self, img: np.ndarray, img_path: pathlib.Path
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:48]: I think it is generally a good practice to indicate the shape and dim definition of the expected img in these data-related functions, e.g. `expects img of shape (C, D, H, W)  where C is ...., D is ....`. Looks like in this case you have (D, H, W) or (C, H, W) image stack (depending on whether you treat the z as a spatial dimension-depth or a channel, i think former is more common).

File: datasets/dataset_00/utils/ImageSelector.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    ) -> dict[pathlib.Path, list[dict[str, Any]]]:
+
+        zslice_groups = defaultdict(list)
+
+        for z_index in range(
+            self.neighbors_per_side,
+            img.shape[0] - self.neighbors_per_side,
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:61]: It also appears that you are using the `img` parameter only for `img.shape[0]`. In my opinion, having some object as input yet only using one of its attributes makes the code unnecessarily confusing.

File: datasets/dataset_00/utils/ImageSelector.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            z_slices_input = self.select_zslices(
+                tifffile.imread(img_path["input_path"]).astype(np.float32)
+                / self.input_max_pixel_value,
+                img_path["input_path"],
+            )
+
+            target_img = tifffile.imread(img_path["target_path"])
+            target_img = (target_img != 0).astype(np.float32)
+
+            z_slices_target = self.select_zslices(
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:199]: With separate calls of `self.select_zslices` to the input and target images and paths are you accounting for potential mismatch in z dimension of the input and target and selecting different z stack slices for the input and target? (also indicated by the logic in `self.filter_symmetric_centered_overlapping_slices`). If so, why would this be useful?  

If that is not the case and we always expect the input image stack and target mask stack to align in dimension (as one would normally expect training a image-image segmentation model), making two separate calls of selecting z slices adds confusion. 

Maybe its me being paranoid but I would do the following refactoring for clarity:

 the `self.select_zslices` as
```
def self.select_zslices(z_dim: int) -> List[List[int]]:
    ... # same logic, just return the collection of all index lists representing all z slice groups corresponding to a given z_dim 
```
and here in `__call__`
```
z_dim = tifffile.imread(img_path["input_path"]).shape[0]
z_groups = self.select_zslices(z_dim)
z_slices_input = { # assemble the nested dict in place
    img_path["input_path"].parent: {
        img_path["input_path"]: group for group in z_groups
    }
}

z_slices_target = { # assemble the nested dict in place
    img_path["target_path"].parent: {
        img_path["target_path"]: group for group in z_groups
    }
}

data_locations.append({"input": z_slices_input, "target": z_slices_target})
```

So the class `__call__` is more self explanatory of the return dictionary structure and each involved helper method's signature accurately reflects what is really needed.

File: datasets/dataset_00/utils/ImageSelector.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        self, img_paths: list[dict[str, pathlib.Path]]
+    ) -> list[dict[str, Any]]:
+
+        self.crop_coords = self.generate_crop_coords()
+        data_locations = []
+
+        for img_path in img_paths:
+            z_slices_input = self.select_zslices(
+                tifffile.imread(img_path["input_path"]).astype(np.float32)
+                / self.input_max_pixel_value,
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:192]: I don't think the `self.select_zslices` uses actual image stack pixel values anywhere, is this normalization left-over from earlier functionality?

File: metrics/BCE.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+        self.reset()
+
+    def reset(self):
+        self.total_loss = torch.tensor(0.0, device=self.device)
+        self.total_elements = 0
+
+    def forward(
+        self,
+        generated_predictions: torch.Tensor,
```
Comment (wli51) [inline metrics/BCE.py:39]: A comment here describing the input tensor shape would be helpful 
e.g. `# (B, D, H, W) logit mask` (correct me if i am understanding the  `generated_predictions` wrong)

File: metrics/ConfusionMetrics.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            self.true_positives / (self.true_positives + self.false_negatives)
+            if (self.true_positives + self.false_negatives) > 0
+            else torch.tensor(0.0, device=self.device)
+        )
+
+        key_precision = f"precision__{self.data_split_logging}"
+        key_recall = f"recall__{self.data_split_logging}"
+
+        self.reset()
+        return {key_precision: precision, key_recall: recall}
```
Comment (wli51) [inline metrics/ConfusionMetrics.py:80]: since you are at it, how about adding yet another metric `f1_score: ...` so you have a single metric that comprises both the precision and recall?

File: models/UNet.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+    ):
+        super().__init__()
+
+        self.inl = Conv(
+            in_channels=in_channels,
+            out_channels=64,
```
Comment (wli51) [inline models/UNet.py:24]: the magic number 64 is usually included as a tunable parameter in UNets and the training speed/VRAM consumption would benefit from lowering this for less intensive/expressive tasks like segmentation (potentially you could pivot the extra memory/compute that is freed to have larger receptive field or batch size).

File: datasets/dataset_00/utils/ImagePreProcessor.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        self.target_transform = target_transform
+
+    def set_image_specs(
+        self, input_max_pixel_value: int, input_ndim: int, target_ndim: int, **kwargs
+    ) -> None:
+        self.input_max_pixel_value = input_max_pixel_value
+        self.input_ndim = input_ndim
+        self.target_ndim = target_ndim
+
+    def format_img(self, img: np.ndarray, img_dims: int) -> torch.Tensor:
```
Comment (wli51) [inline datasets/dataset_00/utils/ImagePreProcessor.py:32]: comments describing the img dimension would be good

File: datasets/dataset_00/utils/ImagePreProcessor.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+        else:
+            raise ValueError(
+                f"The number of dimensions in your image should be 2 or 3. It is currently {img_dims}"
+            )
+
+        return img.to(dtype=torch.float32, device=self.device)
+
+    def __call__(
+        self, input_img: np.ndarray, target_img: np.ndarray = None
```
Comment (wli51) [inline datasets/dataset_00/utils/ImagePreProcessor.py:58]: comments describing the input and target img dimension would be good. Also it looks like you are using Albumentation transforms here. Depending on the transform logic, certain Albumentation transforms like the ones involving rotate and flips require the image numpy ndarray to be strictly (H, W, C) to work properly, so properly describing and handling the dimension can be extra important.

File: datasets/dataset_00/utils/ImagePreProcessor.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Tuple[int, int, int, int]]:
+
+        if self.input_transform is not None:
+            input_img = self.input_transform(image=input_img)["image"]
+
+        if self.target_transform is not None:
+            target_img = self.target_transform(image=target_img)["image"]
+
```
Comment (wli51) [inline datasets/dataset_00/utils/ImagePreProcessor.py:75]: i also find having a input/target shared transform pretty useful for having shared and state matches image augmentation transforms applied the same way to input and target. This can be less relevant for segmentation though.

File: datasets/dataset_00/utils/ImageSelector.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+from typing import Any, Union
+
+import numpy as np
+import tifffile
+import torch
+
+
+class ImageSelector:
+    """
+    Selects the z-slices and crops to be passed as input to the segmentation model.
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:12]: maybe some high level description of the `__call__` return would be good here for clarity

File: datasets/dataset_00/CellSlicetoSliceDataset.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        self.patient_folders = patient_folders
+
+        self.__input_transform = input_transform
+        self.__target_transform = target_transform
+
+        self.data_paths = self.get_image_paths()
+        image_specs = self.get_image_specs()
+        image_selector.set_image_specs(**image_specs)
+        self.data_crops = image_selector(self.data_paths)
+        self.image_preprocessor = image_preprocessor
```
Comment (wli51) [inline datasets/dataset_00/CellSlicetoSliceDataset.py:32]: Maybe i am missing this but how are the `input_transform` and `target_transform` relayed to the `image_preprocessor` object. Seems to me that the preprocessor class has logic for applying the albumentation transforms.

File: callbacks/Callbacks.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+from mlflow.models import ModelSignature
+from mlflow.models.signature import infer_signature
+from torch.amp import autocast
+from torch.nn import Module
+from torch.utils.data import DataLoader
+
+
+class Callbacks:
+    """
+    Triggered at certain points during model training according to the callback hook.
```
Comment (wli51) [inline callbacks/Callbacks.py:14]: Given the multi-purpose nature of this callback class the docstring here could be richer and describe different behavior during different stages (e.g. on epoch start prints a message, on epoch end returns a flag for early termination e.g.)

File: datasets/dataset_00/utils/ImageSelector.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            and len(target_slice) <= len(input_slice)
+        )
+
+    def generate_crop_coords(self):
+        """
+        Naive crop selection
+        """
+
+        crop_coords = []
+        y_starts = list(
```
Comment (wli51) [inline datasets/dataset_00/utils/ImageSelector.py:150]: The assumpution you are making here of all images being of consistent H and W is reasonable. The rigid x,y plane cropping contrasts with the flexible z slicing that allows for potentially shallower target stack (?). I think some docstring justifying for the assymetry in cropping design of xy vs z would be good to have.

File: train.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    raise FileNotFoundError("No Git root directory found.")
+
+# |%%--%%| <5uHY3JoEKK|ExjIoeHzuw>
+r"""°°°
+# Inputs
+°°°"""
+# |%%--%%| <ExjIoeHzuw|uHCF3KHHYz>
+
+root_data_path = root_dir / "big_drive/NF1_organoid_processed_patients"
+patient_folders = [[p for p in root_data_path.iterdir() if root_data_path.is_dir()][0]]
```
Comment (wli51) [inline train.py]: i am puzzled

shouldn't `[[p for p in root_data_path.iterdir() if root_data_path.is_dir()][0]]` be
`[[p for p in root_data_path.iterdir() if p.is_dir()][0]]` instead because otherwise the inner list will include non-dir ps? Also, why [0] after generating a list from iterdir()?

File: train.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+callbacks_args = {
+    "early_stopping_counter_threshold": 3,
+    "image_savers": image_saver,
+    "image_postprocessor": image_postprocessor,
+}
+
+# |%%--%%| <sv6R19116h|1ibSiDMEcz>
+
+unet = UNet(in_channels=1, out_channels=1).to(device)
```
Comment (wli51) [inline train.py]: the `to(device)` seems redundant as the trainer also moves the model

**Review by wli51 — APPROVED** [review]
Great Job! PR LGTM. 
Most of my comments are just nit-picks/request for richer docstring. 
The only things that stood out are:
1. Whether CellSlicetoSliceDataset relays the transforms to ImagePreProcessor.py
2. Unclear `patient_folders` construction in train script

File: datasets/dataset_00/utils/ImageSelector.py (hunk 2 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:61]: I think it can, in this case I was developing with extensibility in mind. For example, I may want to select slices based on other image characteristics, such as if the image does not consist of mostly zero-valued pixels.

File: datasets/dataset_00/utils/ImageSelector.py (hunk 1 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:48]: Good catch, I think this will help with understanding the code.

File: datasets/dataset_00/utils/ImageSelector.py (hunk 3 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:199]: > With separate calls of self.select_zslices to the input and target images and paths are you accounting for potential mismatch in z dimension of the input and target and selecting different z stack slices for the input and target? (also indicated by the logic in self.filter_symmetric_centered_overlapping_slices). If so, why would this be useful?

Keep in mind that this is designed to be more easily modified for future training. Right now I'm framing the image selection in terms of two filtering criteria (one for the input and one for the target). Then I can decide which slices to keep contingent on which slices (if any) satisfy both criteria.

Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:199]: > If that is not the case and we always expect the input image stack and target mask stack to align in dimension (as one would normally expect training a image-image segmentation model), making two separate calls of selecting z slices adds confusion.

I can see your point, however I think the ImageSelector has this responsibility, so I orchestrate the image selection in the `__call__` function of both sets of slices (input and target images). Also, this extensible so that I can later modify class to group the input and target image selection logic.

File: datasets/dataset_00/utils/ImageSelector.py (hunk 4 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:192]: Nope, this is intentional to make the code more extensible (see the comment above)

File: metrics/ConfusionMetrics.py (hunk 6 above)
Comment (MattsonCam) [reply metrics/ConfusionMetrics.py:80]: I think the dice coefficient is the same as the f1 score for binary segmentation. Are there others you think would be beneficial here?

File: metrics/BCE.py (hunk 5 above)
Comment (MattsonCam) [reply metrics/BCE.py:39]: The shape doesn't matter here, as long as `generated_predictions` is the same shape as `targets`

File: models/UNet.py (hunk 7 above)
Comment (MattsonCam) [reply models/UNet.py:24]: I'm planning on just changing the UNet.py code if I want to modify any of the architecture. Also, I'm hoping that AMP speeds this up

File: metrics/ConfusionMetrics.py (hunk 6 above)
Comment (wli51) [reply metrics/ConfusionMetrics.py:80]: I did not realize that the F1 score and Dice are equivalent here. That is nice to know. The only other useful metric I can think of is the Rand Index (RI). To compute it you would need an extra object separation step but I think that could be as simple as grouping connected components. The Rand Index is the ratio of pixels that are assigned to the background and to the same object by both the predicted and the truth, over all the sum of pixels that the truth and predicted has different assignments to. It is a stricter metric and takes into account of pixels assigned to one object vs another that you might benefit from.

Comment (MattsonCam) [reply metrics/ConfusionMetrics.py:80]: I'll keep that in mind

File: train.py (hunk 15 above)
Comment (MattsonCam) [reply train.py]: Good catch, there aren't any files in this directory, so luckily this still worked

File: datasets/dataset_00/utils/ImagePreProcessor.py (hunk 8 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImagePreProcessor.py:32]: Good idea, added

File: datasets/dataset_00/utils/ImagePreProcessor.py (hunk 9 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImagePreProcessor.py:58]: Good point, added

File: datasets/dataset_00/utils/ImagePreProcessor.py (hunk 10 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImagePreProcessor.py:75]: It could be useful, I just see it as a way to add flexibility if needed. I might add more logic to just use one of the transforms if I want each sampled transform to be applied equally to both the input and target

File: datasets/dataset_00/utils/ImageSelector.py (hunk 11 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:12]: Added

File: datasets/dataset_00/CellSlicetoSliceDataset.py (hunk 12 above)
Comment (MattsonCam) [reply datasets/dataset_00/CellSlicetoSliceDataset.py:32]: They are in the init fuction of the `ImagePreProcessor`, I just didn't remove the arguments from the init function of the Dataset class

File: callbacks/Callbacks.py (hunk 13 above)
Comment (MattsonCam) [reply callbacks/Callbacks.py:14]: Most of the functionality is in the _on_epoch_end hook, so I will add more documentation there and mention this point in the class doc-string

File: train.py (hunk 16 above)
Comment (MattsonCam) [reply train.py]: Good catch, I'll change this

File: datasets/dataset_00/utils/ImageSelector.py (hunk 14 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImageSelector.py:150]: This might be changed in the future. Just added a doc-string for this. The idea is that the model can use multiple z-slices of the same cell to improve segmentation performance

**Comment by MattsonCam** [comment]
Thanks for the review @wli51, merging now!
